### PR TITLE
Add dry_run to add()

### LINKED ISSUE #399: feature request: `dry_run` for `add` method
### 🚀 The feature

A dry run method on add could show the chunks. This way you can make sure loader and chunker work as intended.

### Motivation, pitch

This way you have no cost when testing loaders.

### Alternatives

_No response_

### Additional context

It could be implemented in a similar way to `query` and `chat` dry_run.

### PR DESCRIPTION
**Pull request by Dev-Khant**
## Description

Added **dry_run** method to `add` in Embedchain class.

Fixes #399 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update

## How Has This Been Tested?

`poetry run pytest`

- [x] Unit Test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: embedchain/embedchain.py [hunk 1]
```diff
@@ -136,6 +138,9 @@ def add(
         if data_type in {DataType.DOCS_SITE}:
             self.is_docs_site_instance = True
 
+        if dry_run:
+            return documents
```
Comment (cachho) [inline embedchain/embedchain.py:142]: could you make it return a dict of

```python
{
    "chunks": {},
    "count": 0,
    "type": DataType
}
```
I mean count can obviously be deduced, but the type would be quite helpful.

File: docs/advanced/testing.mdx [hunk 2]
```diff
@@ -26,4 +24,15 @@ A: Naval Ravikant is an Indian-American entrepreneur and investor.
...
+
+```python
+print(naval_chat_bot.add('https://navalmanack.s3.amazonaws.com/Eric-Jorgenson_The-Almanack-of-Naval-Ravikant_Final.pdf', dry_run=True))
+
+'''
+['THE ALMANACK OF NAVAL RAVIKANT', 'GETTING RICH IS NOT JUST ABOUT LUCK; HAPPINESS IS NOT JUST A TRAIT WE ARE BORN WITH. These aspirations may seem out of reach, but building wealth and be - ing happy are skills we can learn. So what are these skills, and how do we learn them? What are the principles that should guide our efforts? What does progress really look like? Naval Ravikant is an entrepreneur, philosopher, and investor who has captivated the world with his principles for building wealth and creating long-term happiness. The Almanack of Naval Ravikant is a collection of Naval’s wisdom and experience from the last ten years, shared as a curation of his most insightful interviews and poignant reflections. This isn’t a how-to book, or a step-by-step gimmick. In - stead, through Naval’s own words, you will learn how to walk your own unique path toward a happier, wealthier life. This book has been created as a public service. It is available for free download in pdf and e-reader versions on Navalmanack.com. Naval is']
```
Comment (cachho) [inline docs/advanced/testing.mdx:36]: Could you use a chunker config, and pick a smaller chunking? I think that would be a better example.

Comment (cachho) [inline docs/advanced/testing.mdx:36]: Why is it cut off in the end?

File: embedchain/embedchain.py (hunk 1 above)
Comment (cachho) [reply embedchain/embedchain.py:142]: possibly meta_data too?

**Review by cachho — CHANGES_REQUESTED** [review]
Very nice, I was just thinking about this today, this could've helped a user that needed help. I have a few suggestions to make it better. But solid start, thanks.

**Comment by cachho** [comment]
Can you add a unit test too? You can take `tests/chunkers/test_text.py/test_small_chunksize()` as inspiration and chunk in 1 letter chunks and make sure the size matches and the other properties if you introduce a dict.

File: embedchain/embedchain.py (hunk 1 above)
Comment (Dev-Khant) [reply embedchain/embedchain.py:142]:
```
def load_and_embed(self, loader: BaseLoader, chunker: BaseChunker, src, metadata=None, source_id=None):
        """
        Loads the data from the given URL, chunks it, and adds it to database.

        :param loader: The loader to use to load the data.
        :param chunker: The chunker to use to chunk the data.
        :param src: The data to be handled by the loader. Can be a URL for
        remote sources or local content for local loaders.
        :param metadata: Optional. Metadata associated with the data source.
        :param source_id: Hexadecimal hash of the source.
        :return: (List) documents (embedded text), (List) metadata, (list) ids, (int) number of chunks
        """
        embeddings_data = chunker.create_chunks(loader, src)

        # spread chunking results
        documents = embeddings_data["documents"]
        metadatas = embeddings_data["metadatas"]
        ids = embeddings_data["ids"]
```

If metadata and documents both are needed then we can directly use `embeddings_data` and drop ids from it. And in the same function we can get count which is returned as `count_new_chunks`  and ya data_type can be taken from the func param itself. So let me know if we can use `embeddings_data` directly

File: docs/advanced/testing.mdx (hunk 2 above)
Comment (Dev-Khant) [reply docs/advanced/testing.mdx:36]: Well I too first 2 times from list of documents, will fix it.

Comment (Dev-Khant) [reply docs/advanced/testing.mdx:36]: Yes exactly it will be better

**Comment by Dev-Khant** [comment]
> Can you add a unit test too? You can take `tests/chunkers/test_text.py/test_small_chunksize()` as inspiration and chunk in 1 letter chunks and make sure the size matches and the other properties if you introduce a dict.

Sure I'll look for how it can be done

File: embedchain/embedchain.py (hunk 1 above)
Comment (cachho) [reply embedchain/embedchain.py:142]: wouldn't count `count_new_chunks` actually require for the embeddings to be added (i.e. not a dry run)? Couldn't we just count the list items?

File: docs/advanced/testing.mdx (hunk 2 above)
Comment (cachho) [reply docs/advanced/testing.mdx:36]: you limited it? ok acceptable, maybe add a comment `cut-off for readability`

Comment (Dev-Khant) [reply docs/advanced/testing.mdx:36]: Alright

File: embedchain/embedchain.py (hunk 1 above)
Comment (Dev-Khant) [reply embedchain/embedchain.py:142]: Ok then will use length of documents

Comment (Dev-Khant) [reply embedchain/embedchain.py:142]: And I had one more doubt @cachho.
We need to return `embedding_data` which contains documents and metadatas in `load_and_embed` function before actually pushing it to db right?

**Comment by cachho** [comment]
also please make sure that no telemetry event is fired for this.

File: embedchain/embedchain.py (hunk 1 above)
Comment (cachho) [reply embedchain/embedchain.py:142]: can we add an early return if it's a dry run?

Comment (Dev-Khant) [reply embedchain/embedchain.py:142]: Yes or else it will be pushed into the db and then will get the documents.

Comment (cachho) [reply embedchain/embedchain.py:142]: pushing to the db is not an option. please add an early return.

File: docs/advanced/testing.mdx [hunk 3]
```diff
@@ -33,6 +33,8 @@ The dry run will still consume tokens to embed your query, but it is only **~1/1
...
+{'chunks': ['GETTING RICH IS NOT JUST ABOUT'], 'metadata': [{'source': 'C:\\Users\\Dev\\AppData\\Local\\Temp\\tmp7_cnh810\\tmp.pdf', 'page': 2, 'url': 'https://navalmanack.s3.amazonaws.com/Eric-Jorgenson_The-Almanack-of-Naval-Ravikant_Final.pdf', 'data_type': 'pdf_file'}, 'count': 9940, 'data_type': DataType.PDF_FILE]}
+
+# only one item for readability
```
[the remark this answers is not in the document]
Comment (cachho) [reply docs/advanced/testing.mdx:38]: make the source something simpler like `~/doc.pdf`

[the remark this answers is not in the document]
Comment (cachho) [reply docs/advanced/testing.mdx:38]: and try to prettify the dict

[the remark this answers is not in the document]
Comment (cachho) [reply docs/advanced/testing.mdx:38]: thanks

File: embedchain/embedchain.py [hunk 4]
```diff
@@ -133,13 +133,13 @@ def add(
         data_formatter = DataFormatter(data_type, config)
         self.user_asks.append([source, data_type.value, metadata])
         documents, _metadatas, _ids, new_chunks = self.load_and_embed(
-            data_formatter.loader, data_formatter.chunker, source, metadata, source_id
+            data_formatter.loader, data_formatter.chunker, source, metadata, source_id, dry_run
```
Comment (cachho) [inline embedchain/embedchain.py:136]: because metdatas is used now, you should remove the _underscore.

**Comment by cachho** [comment]
and another thing, instead of just returning the items to be printed, can you also do `logging.debug` on the results? So that users that have debugging log level enabled get them automatically. thanks.

**Comment by Dev-Khant** [comment]
@cachho Made all the required changes. Please do have a look and let me know if any further improvments are required. Thanks

**Comment by Dev-Khant** [comment]
Tests failing because of lint issues in other files from latest changes.

**Review by cachho — APPROVED** [review]
looks good

**Review by taranjeet — APPROVED** [review]
looks good.

**Comment by Dev-Khant** [comment]
Failed tests are fixed.

**Comment by Dev-Khant** [comment]
It's ready @taranjeet. Let me know if there are any more improvements. Thanks

**Pull request merged**